Web Docs - Stepper Component (#2727)
hashicorp/design-system: website/docs/components/stepper/nav/partials/guidelines/guidelines.md

Reviewer: The heading order on this page could be reorganized based on importance and aligned with common formatting from other pages. 

Maybe consider something like this hierarchy instead:

- Usage
- Interactive vs non-interactive
- Content
- Spacing
- Mirroring experiences (should move to "Composition with other components")

## Usage

### When to use

- In complex flows that are broken up into multiple steps to more easily segment content.

> - To communicate progress in a mult-step flow; what has been completed, and what has yet to be completed.

Reviewer: removing repetition

```suggestion
- To communicate progress in a mult-step flow.
```

Or something like:
"To communicate what has and has not been completed in a mult-step flow."

- For a single, on-page experience that requires completion.

### When not to use 

> - Non-complex flow, use the [form pattern](/patterns/form-patterns) on a single page.

Reviewer: ```suggestion
- For non-complex flows, use the [form pattern](/patterns/form-patterns) on a single page.
```

> - Don't require specific progress tracking, use structured page content.

Reviewer: ```suggestion
- For simple forms or when progress doesn't need to be tracked, use structured page content.
```

- For progression that doesn’t require a user to remain on the page for, use the [Stepper List](/components/stepper/list).

## Content

> Content in the Stepper should be brief and to the point. The Stepper has a title and an optional description.

Reviewer: ```suggestion
The Stepper Nav has a title and an optional description. Content should be brief and to the point.
```

### Title

> Titles are required and help users understand, from an overview perspective, what tasks need to be completed. Titles should be brief, and only a couple of words.

Reviewer: [suggestion]
```suggestion
Titles provide an overview to help users understand what tasks need to be completed. Titles are required and should be brief.
```

### Description

> Descriptions are optional and should be brief and provide additional context to the title. We recommend that all steps either have a description or none do.

Reviewer: [suggestion]
```suggestion
Descriptions are optional. They should provide additional context but be brief. We recommend that all steps either have a description or do not.
```

!!! Do

![The stepper with descriptions enabled in every step.](/assets/components/stepper/navigation/stepper-navigation-description-do.png)

!!!

> !!! Dont

Reviewer: We typically try to add a short sentence explaining what the image is showing in do/dont blocks.

```suggestion
!!! Dont

Don't mix and match description visibility.
```

![Steps one and four with enabled descriptions, while two and three have only a title.](/assets/components/stepper/navigation/stepper-navigation-description-dont.png)

!!!

### Number of steps

> It is recommended that the max number of steps be 4, however if that isn’t possible, be mindful of how the content within the title and description wraps in smaller screens.

Reviewer: Is this recommendation purely based on size limitations, or is there another reason? If there's another reason, can we provide more helpful details as to why?

```suggestion
We recommended the maximum number of steps be 4. If that isn’t possible, be mindful of how the content within the title and description wraps on smaller screens.
```

Author: It's size limitations (even when we wrap this component responsively).

Reviewer: I'm curious to hear what others think, but I, personally, don't think this is a good enough reason to recommend a limited number of steps. There are likely other valid reasons for reducing the number of steps, though, so if we have thoughts on that, we should include those. If not, we should reevaluate including this as a recommendation.

> ## Interactive vs non-interactive

Reviewer: There may be a more scannable hierarchy. Something like:

- Interactivity
  - Interactive steps
  - Non-interactive steps

Where the details go into each relevant section. This is also more scalable, so details can be added later by variant type.

> The Stepper Nav allows for both interactive (default) and non-interactive variants. Once a user navigates backward with the interactive variant, they cannot progress forward through the Stepper. The Non-interactive variant is for read-only purposes and does not allow backwards navigation within the Stepper.

Reviewer: As mentioned above, I suggest breaking this up by variant type; interactive and non-interactive.

Reviewer: > Once a user navigates backward with the interactive variant, they cannot progress forward through the Stepper.

We should do a bit more explaining about why this is. Given that it's already been a hot topic within our team, I'm sure it's something we'll get questions about from consumers.

> ![Subtitle "Interactive" followed by an interactive variant of the Stepper Nav. Another subtitle "Non-interactive" followed by a non-interactive variant of the Stepper Nav.](/assets/components/stepper/navigation/stepper-navigation-interactive-non-interactive.png)

Reviewer: I'd separate these two examples so you can show both interactive examples near one another.

The interactive variant allows for backward movement after a step has been completed.

![The user's mouse is hovering over step one, showing its hover state.](/assets/components/stepper/navigation/stepper-navigation-interactive-backwards-movement.png)

## Spacing

> To account for the absolutely positioned indicator within each Step, it is recommended to provide a 32px padding on the top of the Stepper and 24px below. 

Reviewer: This is a consistency thing, so I don't think it needs to be a "recommendation". The absolute positioning issue only feels relevant for Figma and I'm not sure it's needed in the docs. Curious what others think though. 

```suggestion
There should be 32px of padding above the Stepper Nav and 24px below. 
```

![](/assets/components/stepper/navigation/stepper-navigation-spacing-example.png)

> ## Mirroring experiences

Reviewer: Typically, we put this type of content under a heading called "Composition with other components" ([ref](https://hds-website-git-stepper-docs-hashicorp.vercel.app/components/copy/button#composition-with-other-components))

```suggestion
## Composition with other components

### With button sets
```

> A [Button Set](/components/button-set) should always be present alongside a Stepper. Ensure that navigation between the Stepper and the Button Set matches. If a user can navigate backward within the Button Set, use the interactive variant. Otherwise, use the non-interactive variant. It is recommended that the Stepper’s width matches the width of the page’s content.

Reviewer: ```suggestion
Always pair a Stepper Nav with a [Button Set](/components/button-set). Ensure that navigation interactions between the Stepper Nav and the Button Set match. 

If a user can navigate backward within the Button Set, use the interactive variant. Otherwise, use the non-interactive variant.
```

> A [Button Set](/components/button-set) should always be present alongside a Stepper. Ensure that navigation between the Stepper and the Button Set matches. If a user can navigate backward within the Button Set, use the interactive variant. Otherwise, use the non-interactive variant. It is recommended that the Stepper’s width matches the width of the page’s content.

Reviewer: > It is recommended that the Stepper’s width matches the width of the page’s content.

This doesn't feel like it belongs in this section. I might consider moving it to its own section near the Spacing one that talks about width and resizing options.

> ![](/assets/components/stepper/navigation/stepper-navigation-matching-experiences.png)

Reviewer: The sticky footer in this example feels a bit like it's floating. Could the space be tightened up a little so it's clear they're a part of the same example/screenshot?

## Usage

### When to use

> - In complex flows that are broken up into multiple steps to more easily segment content.

Reviewer 2: ```suggestion
- In complex flows that are broken up into multiple steps.
```

Reviewer 2: the last part seems a bit redundant to me

- To communicate what has and has not been completed in a multi-step flow.

> - For a single, on-page experience that requires completion.

Author: ```suggestion
- For a single page multi-step flow that requires completion.
```

> - For a single, on-page experience that requires completion.

Reviewer 2: What is a single, on-page experience? Is there a real world use case example to clarify?

Author: One single page, where completing the stepper requires you to remain on the page.

Reviewer 2: Thanks, how about something more like this then?

"For a single page multi-step flow vs a multi-page flow."

Author: Added below.

Reviewer 3: Would it be clearer if it was something like.... "For experiences that must be completed in one session and does not have steps that link out to tasks on other pages." ?

### When not to use

> - For non-complex flows, use the [form pattern](/patterns/form-patterns) on a single page.

Reviewer 2: ```suggestion
- For simple flows, use the [form pattern](/patterns/form-patterns) on a single page.
```

- For simple forms or when progress doesn't need to be tracked, use structured page content.

> - For progression that doesn’t require a user to remain on the page for, use the [Stepper List](/components/stepper/list).

Reviewer 2: ```suggestion
- For a progression of tasks that doesn’t require a user to remain on the same page, use the [Stepper List](/components/stepper/list).
```

## Interactive vs non-interactive

> The Stepper Nav supports both interactive (default) and non-interactive variants.

Reviewer 4: [Issue] We may need to discuss this, but currently the nav is **non-interactive** by default.

Author: Do we believe our consumers would use the non-interactive variant more often than the alternative? If so, we can make this update in my Figma component and here to:

```suggestion
The Stepper Nav supports both non-interactive (default) and interactive variants.
```

Author: This is currently based off of the [requirements doc in the variants](https://docs.google.com/document/d/1itgRQ8qmsMRAld_-GRkixlp25ud-WWk6K3viXaJvHyg/edit?tab=t.efihpzui45kb) section indicating that `isInteractive` set to `true` is the default.

Do we believe that our consumers will use the non-interactive variant more often than the alternative? That's usually how we define what is set by default.

Author: Discussed w/ Dylan and we are setting the `isInteractive = true` as the default

> The Stepper Nav allows for both interactive (default) and non-interactive variants. 

Reviewer 5: [non-blocking suggestion]

```suggestion
The Stepper Nav supports both interactive (default) and non-interactive variants. 
```

### Interactive steps

> A user is able to navigate backwards within the Stepper Nav, however, they cannot progress forward. The reason for this is to ensure consistency in behavior and no dirty forms exist.

Reviewer 6: ```suggestion
A user is able to navigate backwards within the Stepper Nav, however, they cannot navigate forward unless they have completed their current step. This creates consistency for step flow behavior and order of operations.
```

This reads strangely, not sure if my suggestion above is clearer but something in that direction would maybe be good.

Should we also add something like — Information filled in a step will persist if a user navigates backward?

Author: I don't think we should, this depends on the consumer's implementation. Often times it will clear out.

> A user can navigate backwards within the Stepper Nav steps but cannot navigate forwards unless they have completed the current step. This creates consistency for step flow behavior and the order of operations.

Reviewer 2: ```suggestion
Users can navigate backwards within the Stepper Nav steps but cannot navigate forwards unless they have completed the current step. This creates consistency for step flow behavior and the order of operations.
```

> A user is able to navigate backwards within the Stepper Nav, however, they cannot navigate forward unless they have completed the current step. This creates consistency for step flow behavior and order of operations.

Reviewer 2: ```suggestion
A user can navigate backwards within the Stepper Nav steps but cannot navigate forwards unless they have completed the current step. This creates consistency for step flow behavior and the order of operations.
```

> A user is able to navigate backwards within the Stepper Nav, however, they cannot navigate forward unless they have completed their current step. This creates consistency for step flow behavior and order of operations.

Reviewer 5: [suggestion] "steps" aren't really possessive in this case I suppose.

```suggestion
A user is able to navigate backwards within the Stepper Nav, however, they cannot navigate forward unless they have completed the current step. This creates consistency for step flow behavior and order of operations.
```

![](/assets/components/stepper/navigation/stepper-navigation-interactive-steps.png)

![The user's mouse is hovering over step one of four, showing its hover state.](/assets/components/stepper/navigation/stepper-navigation-interactive-backwards-movement.png)

### Non-interactive steps

> The non-interactive variant is for read-only purposes and does not allow backwards navigation within the Stepper Nav.

Reviewer 6: ```suggestion
The non-interactive variant is for read-only purposes and does not allow backwards navigation.
```

![](/assets/components/stepper/navigation/stepper-navigation-non-interactive-steps.png)

## Content

> The Stepper Nav has a title and an optional description. Content should be brief and to the point.

Reviewer 6: ```suggestion
The Stepper Nav has a title and an optional description. Content should limited to necessary context, long form content should live within the body of the page.
```

Author: Updating yours a little to fix a sentence issue:
```suggestion
The Stepper Nav has a title and an optional description. Content should be limited to necessary context, long form content should live within the body of the page.
```

> The Stepper Nav supports a title and an optional description. Content should be limited to only what is necessary, and long form content should live within the body of the page.

Reviewer 3: ```suggestion
The Stepper Nav supports a title and an optional description. Content should be limited to only what is necessary, and long form content should live within the body of the Step.
```

Just to align the content with the code API a bit more.

> The Stepper Nav has a title and an optional description. Content should be limited to necessary context, and long form content should live within the body of the page.

Reviewer 5: [suggestion] Somewhat more formal, and removal of "context" which read a bit strange to me.

```suggestion
The Stepper Nav supports a title and an optional description. Content should be limited to only what is necessary, and long form content should live within the body of the page.
```

### Title

Titles provide an overview to help users understand what tasks need to be completed. Titles are required and should be brief.

### Description

> Descriptions are optional. They should provide additional context but be brief. We recommend that all steps either have a description or do not.

Reviewer 6: "brief" is used a ton in this page, maybe it can be swapped for something else in a few places?

!!! Do

> Ensure description visibility is consistent.

Reviewer 6: "Ensure" is also used a lot.
```suggestion
Implement consistent description visibility.
```

> Descriptions are optional. They should provide additional context but be brief. We recommend that all steps either have a description or do not.

Reviewer 5: I might recommend flipping this sentence because you lead with the value prop and the required/not required point in the Title section.

```suggestion
Descriptions provide additional context about the step and are optional. We recommend that all steps either have a description or do not.
```

> Implement consistent description visibility.

Reviewer 2: ```suggestion
Use descriptions consistently.
```

Reviewer 2: "visibility" doesn't seem like the right word here as, to me, it implies that there may be invisible or hidden descriptions.

> ![The stepper with descriptions enabled in every step.](/assets/components/stepper/nav/stepper-nav-description-do.png)

Reviewer 2: ```suggestion
![The stepper with descriptions included in every step.](/assets/components/stepper/nav/stepper-nav-description-do.png)
```

!!!

!!! Dont

> Don't mix and match description visibility.

Reviewer 2: ```suggestion
Don't mix and match inclusion of descriptions.
```

> ![Steps one and four include descriptions, while steps two and three have only a title.](/assets/components/stepper/nav/stepper-nav-description-dont.png)

Reviewer 3: ```suggestion
![Stepper where steps one and four include descriptions, while steps two and three have only a title.](/assets/components/stepper/nav/stepper-nav-description-dont.png)
```

> ![Steps one and four with enabled descriptions, while two and three have only a title.](/assets/components/stepper/nav/stepper-nav-description-dont.png)

Reviewer 2: ```suggestion
![Steps one and four include descriptions, while steps two and three have only a title.](/assets/components/stepper/nav/stepper-nav-description-dont.png)
```

!!!

### Number of steps

> We recommended the maximum number of steps be 4. If that isn’t possible, be mindful of how the content within the title and description will wrap on smaller screens.

Reviewer 3: ```suggestion
We recommend a maximum of 4 steps. If that isn’t possible, be mindful of how the content within the title and description will wrap on smaller screens.
```

> We recommended the maximum number of steps be 4. If that isn’t possible, be mindful of how the content within the title and description wraps on smaller screens.

Reviewer 5: [non-blocking suggestion]

```suggestion
We recommended the maximum number of steps be 4. If that isn’t possible, be mindful of how the content within the title and description will wrap on smaller screens.
```

## Spacing

> There should be 32px of padding above the Stepper Nav and 24px below.

Reviewer 2: ```suggestion
There should be 32px of spacing above the Stepper Nav and 24px below.
```

Reviewer 2: Changed to the more generic word "spacing" as technically, I think margins should normally be used to add space between elements.

![](/assets/components/stepper/nav/stepper-nav-spacing-example.png)

## Resizing behavior

> Each Step will scale evenly within the Stepper Nav's width. This creates consistency in the sizes of touch targets for hover, active and focus states, especially in larger screens. The total width of the Stepper Nav should match the content of the page.

Reviewer 2: ```suggestion
Each step scales evenly to fit within the total width of the Stepper Nav. This ensures consistent touch target sizes for hover, active, and focus states. The total width of the Stepper Nav should match the page content width.
```

> Each Step will scale evenly within the Stepper Nav's width. This ensures consistency in hit targets for hover, active and focus states, especially in larger screens. Ensure that the Stepper Nav's total width matches the content of the page.

Reviewer 6: ensures twice here

> Each Step will scale evenly within the Stepper Nav's width. This ensures consistency in hit targets for hover, active and focus states, especially in larger screens. Ensure that the Stepper Nav's total width matches the content of the page.

Reviewer 6: ```suggestion
Each Step will scale evenly within the Stepper Nav's width. This creates consistency in hit targets for hover, active and focus states, especially in larger screens. The total width of the Stepper Nav should match the content of the page.
```

> ### Responsive wrapping

Reviewer 2: ```suggestion
### Responsive layout
```

> Each Step will scale evenly within the Stepper Nav's width. This creates consistency in hit targets for hover, active and focus states, especially in larger screens. The total width of the Stepper Nav should match the content of the page.

Reviewer 5: [suggestion] I think "touch targets" is probably a bit more accurate.

Regarding the width: I know what you're trying to say, but I don't think this is saying that. I thought we landed on saying the Stepper should fill the width of the page, which the content on the page isn't always the width of the page (e.g., a form or something like that where the width of elements is intentionally limited). Maybe it's as simple as saying: "The Stepper Nav should fill the available width of the page."

```suggestion
Each Step will scale evenly within the Stepper Nav's width. This creates consistency in the sizes of touch targets for hover, active and focus states, especially in larger screens. The total width of the Stepper Nav should match the content of the page.
```

> At browser widths of 550px and below the steps will stack vertically and not display a connecting progress bar.

Reviewer 4: ```suggestion
At browser widths of 550px and below, the steps will stack vertically and not display a connecting progress bar.
```

> Starting at 550px browser width, the Steps within the Stepper Nav will begin to wrap and the progress bar will be removed.

Reviewer 2: ```suggestion
At browser widths of 550px and below the steps will stack vertically and not display a connecting progress bar.
```

Reviewer 2: The steps don't wrap, they change orientation from horizontal to vertical.

> Once the browser width is 550px, the Steps within the Stepper Nav will wrap, and the progress bar will be removed.

Reviewer 6: ```suggestion
Starting at 550px browser width, the Steps within the Stepper Nav will begin to wrap and the progress bar will be removed.
```

> ![The Stepper Nav is stacked vertically](/assets/components/stepper/nav/stepper-nav-responsive-behavior.png)

Reviewer 3: ```suggestion
![On small screens, the steps in a Stepper Nav switch from horizontal orientation to vertical.](/assets/components/stepper/nav/stepper-nav-responsive-behavior.png)
```

## Composition with other components

> Always pair a Stepper Nav with a [Button Set](/components/button-set). Navigation interactions between the Stepper Nav and the Button Set should always match. 

Author: @KristinLBradley maybe something like this? Thoughts?

```suggestion
Always pair a Stepper Nav with a [Button Set](/components/button-set) or [Button](/components/button). Navigation interactions between the Stepper Nav and the Button Set should always match. 
```

Reviewer 2: That's helpful since use of a ButtonSet is just for the layout while the Buttons are what provide actual functionality but I still think we need to explain everything more clearly. I'll try to look at this again tomorrow and come up with some ideas but I was confused myself by what was meant in this section so I'll have to think about it some more.

Author: ok, I appreciate your input :)

Reviewer 3: How is this?

```suggestion
The Stepper Nav is meant to be used with a [Button Set](/components/button-set) and [Buttons](/components/button) placed at the end of the step content. The buttons allow the user to complete each step and move to the next and optionally go back to a previous step.

If the user is able to move back to previous steps, include a previous button and use the interactive Stepper Nav. Otherwise, use only a primary button to move to the next step and the non-interactive Stepper Nav. This way, the navigation interactions between the buttons and the Stepper Nav always match.
```

Author: The bottom already talks about using the interactive variant vs not. I'll just use mine to keep brief.

> Always pair a Stepper Nav with a [Button Set](/components/button-set). Ensure that navigation interactions between the Stepper Nav and the Button Set match. 

Reviewer 6: ```suggestion
Always pair a Stepper Nav with a [Button Set](/components/button-set). Navigation interactions between the Stepper Nav and the Button Set should always match. 
```

> Always pair a Stepper Nav with a [Button Set](/components/button-set) or [Button](/components/button). Navigation interactions between the Stepper Nav and the Button Set should always match. 

Reviewer 2: ```suggestion
Always pair the Stepper Nav with appropriate navigation [Button](/components/button) such as "Next" and "Previous". Buttons can be wrapped with a [Button Set](/components/button-set) to ensure consistent spacing between them.
```

> Always pair the Stepper Nav with appropriate navigation [Button](/components/button) such as "Next" and "Previous". Buttons can be wrapped with a [Button Set](/components/button-set) to ensure consistent spacing between them.

Reviewer 2: ```suggestion
Always pair the Stepper Nav with appropriate navigation [Buttons](/components/button) such as "Next" and "Previous". Buttons can be wrapped with a [Button Set](/components/button-set) to ensure consistent spacing between them.
```

> If a user can navigate backward within the Button Set, use the interactive variant. Otherwise, use the non-interactive variant.

Reviewer 3: ```suggestion
If a user can navigate backward within the Button Set, use the interactive Stepper Nav. Otherwise, use the non-interactive variant.
```

> If a user can navigate backward within the Button Set, use the interactive variant. Otherwise, use the non-interactive variant.

Reviewer 2: The above text is confusing to me. Are the buttons supposed to be inside of each step? Or should there be buttons outside the Stepper Nav which persist?

What type of buttons should be used? Is this referring to "Previous", "Next", etc. buttons for navigation? (So are the buttons meant to be an alternate way of navigating through the steps vs. clicking on the steps themselves?)

Looking at the Showcase, it seems that there are buttons for navigating between steps which are outside of the Stepper but I wouldn't have grasped that just from the documentation.

Reviewer 4: Not sure what the design guidance is. Will defer to @majedelass. Just a note in the showcase there are buttons inside each panel. 
```
<Hds::Stepper::Nav as |S|>
  <S.Step>
    <:title>Step 1</:title>
  </S.Step>
  <S.Panel>
    <button>Next</button>
  </S.Panel>
</Hds::Stepper::Nav>
```

Reviewer 2: The inclusion of the Buttons/ButtonSet relates directly to the functionality so if you could add suggestions clarifying this usage I think it would be helpful. @dchyun 

@majedelass and other designers can also make their own additional edits & suggestions if they feel it's needed afterwards if you were to make your own.

I feel that writing documentation often works best as a collaborative process between designers & devs. Designers should also feel free to make suggested edits to the "Code" tab for example.

Reviewer 2: It's also not clear whether you should include a ButtonSet & Buttons only with an interactive Stepper? Since this current text just says "Always pair a Stepper Nav with a Button Set".

Author: Maybe it should be Button Set and Buttons to your point, because if they are using the non-interactive variant, then a button set may not be used at all.

Reviewer 2: Well do we want them to always use a ButtonSet with Buttons together with the StepperNav? Or only with the interactive StepperNav?

And what type of Buttons should they use it with and for what purpose? I think we need to be clearer about those details and more explicit to avoid consumers misunderstanding.

Reviewer 2: I submitted suggested edits for the 2 sentences in this section. So please check those out & apply if you agree.

> If a user can navigate backward within the Button Set, use the interactive Stepper Nav. Otherwise, use the non-interactive variant.

Reviewer 2: ```suggestion
If backwards navigation is allowed, use an interactive Stepper Nav, otherwise use the non-interactive variant.
```

> ![](/assets/components/stepper/navigation/stepper-navigation-matching-experiences.png)

Reviewer 5: I think the Buttons in this example are outdated or the sizing is off for some reason (should be square I think?) I left a comment on this in Figma.
